src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -6355,6 +6408,416 @@ register_handler(int sig, struct sigaction *old_handler)
+			return (-EINVAL);
+		}
+	}
+
+	aio_eq_count_local = aio_eq_count;
+	for (i = 0; i < nr; i++) {
+		op         = ios[i]->aio_lio_opcode;
+		fd         = ios[i]->aio_fildes - FD_FILE_BASE;
+		idx_aio_eq = (aio_ctx_obj->first_eq + i) % aio_eq_count_local;
+
+		D_ALLOC_PTR(ctx_ev);
+		if (ctx_ev == NULL) {
+			if (i == 0)
+				return -ENOMEM;
+			else
+				return i;
+		}
+
+		rc = daos_event_init(&ctx_ev->ev, aio_eq_list[idx_aio_eq].eq, NULL);
+		if (rc) {
+			D_FREE(ctx_ev);
+			DL_ERROR(rc, "daos_event_init() failed");
+			D_GOTO(err, rc = daos_der2errno(rc));
+		}
+		d_iov_set(&iov, (void *)ios[i]->u.c.buf, ios[i]->u.c.nbytes);
+		sgl.sg_nr     = 1;
+		sgl.sg_iovs   = &iov;
+		ctx_ev->piocb = ios[i];
+		/* EQs are shared by contexts. Need to save ctx when polling EQs. */
+		ctx_ev->ctx = aio_ctx_obj;
+		if (op == IO_CMD_PREAD) {
+			rc = dfs_read(file_list[fd]->dfs_mt->dfs, file_list[fd]->file, &sgl,
+				      ios[i]->u.c.offset, &read_size, &ctx_ev->ev);
+			if (rc)
+				D_GOTO(err, rc);
+		}
+		if (op == IO_CMD_PWRITE) {
+			rc = dfs_write(file_list[fd]->dfs_mt->dfs, file_list[fd]->file, &sgl,
+				       ios[i]->u.c.offset, &ctx_ev->ev);
+			if (rc)
+				D_GOTO(err, rc);
+		}
+		atomic_fetch_add_relaxed(&aio_ctx_obj->num_op_submitted, 1);
+		atomic_fetch_add_relaxed(&aio_eq_list[idx_aio_eq].n_op_queued, 1);
+	}
+
+	return nr;
+
+err:
+	D_FREE(ctx_ev);
+	return (-rc);
+}
+
+int
+io_cancel(io_context_t ctx, struct iocb *iocb, struct io_event *evt)
+{
+	aio_ctx_t   *aio_ctx_obj = (aio_ctx_t *)ctx;
+	io_context_t ctx_real    = aio_ctx_obj->ctx;
+
+	if (next_io_cancel == NULL) {
+		next_io_cancel = dlsym(RTLD_NEXT, "io_cancel");
+		D_ASSERT(next_io_cancel != NULL);
+	}
+	if (!hook_enabled)
+		return next_io_cancel(ctx_real, iocb, evt);
+
+	if (iocb->aio_fildes < FD_FILE_BASE)
+		return next_io_cancel(ctx_real, iocb, evt);
+
+	/* daos_event_abort() may be used to implement this feature later. */
+	DS_ERROR(ENOSYS, "io_cancel() for DFS is not implemented");
+	return (-ENOSYS);
+}
+
+static int
+ev_enqueue(struct aio_ctx *ctx, struct aio_ev *ev)
+{
+	D_MUTEX_LOCK(&ctx->lock);
+	if (ctx->ev_queue_len >= ctx->depth) {
+		D_MUTEX_UNLOCK(&ctx->lock);
+		/* Unexpected here */
+		assert(0);
+	}
+	ctx->ev_done_list[ctx->ev_done_t] = ev;
+	ctx->ev_done_t++;
+	if (ctx->ev_done_t >= ctx->depth)
+		ctx->ev_done_t = 0;
+	ctx->ev_queue_len++;
+	D_MUTEX_UNLOCK(&ctx->lock);
+	return 0;
+}
+
+static void
+ev_dequeue_batch(struct aio_ctx *ctx, long min_nr, long nr, struct io_event *events, int *num_ev)
+{
+	struct aio_ev *ev;
+
+	D_MUTEX_LOCK(&ctx->lock);
+	while (ctx->ev_queue_len > 0 && *num_ev < min_nr) {
+		/* dequeue one ev record */
+		ev = ctx->ev_done_list[ctx->ev_done_h];
+		ctx->ev_done_h++;
+		if (ctx->ev_done_h >= ctx->depth)
+			ctx->ev_done_h = 0;
+		ctx->ev_queue_len--;
+
+		events[*num_ev].obj = ev->piocb;
+		events[*num_ev].res = ev->piocb->u.c.nbytes;
+		daos_event_fini(&ev->ev);
```

Comment:
rc check

Reply:
Added the check of returned value for daos_event_fini(). Thank you!